src/libraries/Common/src/SourceGenerators/ImmutableEquatableArray.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics.Hashing;
+
+namespace SourceGenerators
+{
+    /// <summary>
+    /// Provides an immutable list implementation which implements sequence equality.
+    /// </summary>
+    public sealed class ImmutableEquatableArray<T> : IEquatable<ImmutableEquatableArray<T>>, IReadOnlyList<T>
+        where T : IEquatable<T>
+    {
+        public static ImmutableEquatableArray<T> Empty { get; } = new ImmutableEquatableArray<T>(Array.Empty<T>());
+
+        private readonly T[] _values;
+        public T this[int index] => _values[index];
+        public int Count => _values.Length;
+
+        public ImmutableEquatableArray(IEnumerable<T> values)
+            => _values = values.ToArray();
```

Comment:
Did you consider just making the parameter type here `T[]` so that consumer can potentially avoid an allocation?

Reply:
It makes a defensive copy to avoid encapsulating a potentially mutable source enumerable.

Reply:
Correct but if you make the parameter `T[]` there is no mutable source enumerable. The caller is forced to allocate the array and in the case the collection is already an array we don't double alloc.

Reply:
Sure, but the array passed in by the caller could theoretically be mutated (e.g. it could be passing a rented buffer). Not super likely to happen here though.

Reply:
if this is all internal, i would make this just a T[].  You know your creators and can ensure they're doing the right thing.